### PR TITLE
Use sourcemaps to revert stacktrace in extension

### LINKED ISSUE #972: Create and ship source map for better extension stack traces
With the new "report error" link in the extension, we've already started getting quite a few more [bugs reported](https://github.com/GoogleChrome/lighthouse/issues/969). It would be great if we could do better on the stack traces, though, and have the specific file instead of the browserified `lighthouse-background.js:5535:17` or whatever.

We may want to intentionally cut out library code if the size is too big, especially the traceviewer/devtools deps, or the devtools version can simply not ship the source map, as size in the extension is somewhat less of a concern.

### PR DESCRIPTION
**Pull request by wardpeet**
Should fix #972, downside is that I had to fix an issue with browserify sourcemaps inside https://github.com/novocaine/sourcemapped-stacktrace/pull/19

perhaps you know of another library? (didn't want to write it myself)

### CODE REVIEW

File: lighthouse-extension/app/src/popup.js [hunk 1]
```diff
@@ -157,7 +166,10 @@ document.addEventListener('DOMContentLoaded', _ => {
 
       if (!multipleTabs && !debuggerExists) {
         feedbackEl.className = 'feedback-error';
-        feedbackEl.appendChild(buildReportErrorLink(err));
+        buildReportErrorLink(err)
```
Comment (brendankenny) [inline lighthouse-extension/app/src/popup.js]: ideally we'd want the error passed to `background.console.error` below to also be source mapped

File: lighthouse-extension/app/src/popup.js [hunk 2]
```diff
@@ -50,16 +52,23 @@ document.addEventListener('DOMContentLoaded', _ => {
...
     qsBody += '**Chrome Version**: ' + getChromeVersion() + '\n';
     qsBody += '**Error Message**: ' + err.message + '\n';
-    qsBody += '**Stack Trace**:\n ```' + err.stack + '```';
 
-    const base = 'https://github.com/googlechrome/lighthouse/issues/new?';
-    const title = encodeURI('title=Lighthouse Extension Error');
-    const body = '&body=' + encodeURI(qsBody);
+    return new Promise(resolve => {
+      sourcemapStacktrace.mapStackTrace(err.stack, function(mapped) {
+        qsBody += '**Stack Trace**:\n ```' + mapped.join('\n') + '```';
```
Comment (brendankenny) [inline lighthouse-extension/app/src/popup.js]: will need to rebase this after #992

File: lighthouse-extension/package.json [hunk 3]
```diff
@@ -23,5 +23,8 @@
     "gulp-zip": "^3.2.0",
     "run-sequence": "^1.1.5",
     "through2": "^2.0.1"
+  },
+  "dependencies": {
+    "sourcemapped-stacktrace": "^1.1.3"
```
Comment (wardpeet) [inline lighthouse-extension/package.json]: i require it inside browserify doesn't that make it a dependency?

File: lighthouse-extension/app/src/popup.js (hunk 1 above)
Comment (wardpeet) [inline lighthouse-extension/app/src/popup.js]: done :)

File: lighthouse-extension/app/src/popup.js (hunk 2 above)
Comment (wardpeet) [inline lighthouse-extension/app/src/popup.js]: done :)

File: lighthouse-extension/package.json (hunk 3 above)
Comment (paulirish) [inline lighthouse-extension/package.json]: i think in practical terms it doesnt matter, but i agree that semantically it fits in deps more than devDeps

File: lighthouse-extension/package.json [hunk 4]
```diff
@@ -23,5 +23,8 @@
     "gulp-zip": "^3.2.0",
     "run-sequence": "^1.1.5",
     "through2": "^2.0.1"
+  },
+  "dependencies": {
```
Comment (paulirish) [inline lighthouse-extension/package.json]: do we need to wait for your PR to land upstream first?

File: lighthouse-extension/app/src/popup.js [hunk 5]
```diff
@@ -122,6 +126,21 @@ document.addEventListener('DOMContentLoaded', _ => {
     return listItem;
   }
 
+  function showError(err, message, includeReportLink) {
+    sourcemapStacktrace.mapStackTrace(err.stack, stackWithSourcemaps => {
```
Comment (paulirish) [inline lighthouse-extension/app/src/popup.js]: can we do a try/catch around the mapStackTrace call? I want to make sure we gracefully degrade and still show the error reporting stuff even if the sourcemap resolution fails.

File: lighthouse-extension/package.json (hunk 4 above)
Comment (wardpeet) [inline lighthouse-extension/package.json]: if necessary I can publish the github repo under my name in npm repo until the pr is accepted

File: lighthouse-extension/app/src/popup.js (hunk 5 above)
Comment (wardpeet) [inline lighthouse-extension/app/src/popup.js]: Need to test my new approach :)

File: lighthouse-extension/package.json (hunk 4 above)
Comment (wardpeet) [inline lighthouse-extension/package.json]: @paulirish PR got accepted ^^

**Comment by wardpeet** [comment]
Background.js looks like this
image

Issue looks like this
image

**Comment by paulirish** [comment]
Tried it out when I was hitting some bugs. Some examples of this PR in action:


##### before and after - exception 1
image
image



##### before and after - exception 2
image
image

**Comment by paulirish** [comment]
I like the added benefit this PR gives our error stack traces. The impact on our bundle size for adding this new module is ~90kb which is similar to the jszip that we removed. 

However on the other hand, this means including sourcemaps for our entire bundle which takes the filesize from 2.4mb to 6.5mb.

**Comment by wardpeet** [comment]
mmh I don't think we can fix the sourcemap issue but I might be able to pick relevant pieces of the library and put them in lighthouse but probably won't save that much because I will still be dependent on the sourcemap dependency

**Comment by wardpeet** [comment]
@paulirish don't forget we are also removing url and whatwg-url from the extension.

**Comment by wardpeet** [comment]
@paulirish I did a new check with a rebase of master and got the following results
Master is still 2.4MB
My branch is 6,0MB

Is it that bad? We could also do a better issue tracking and use the report viewer to convert stacktraces so the extension stays 2MB and the heavy lifting is done by the viewer but than we need to build a small ui for it.

**Comment by brendankenny** [comment]
should we close this? Looks like the impact on code size would be too great? It seems like it hasn't been *too* hard to backtrack from extension strack traces to original source code

**Comment by wardpeet** [comment]
@brendankenny yeah think closing is ok :)

**Pull request closed without merging**